kotsadm/pkg/redact/redact.go

```diff
@@ -132,10 +173,278 @@ func SetRedactSpec(spec string) (string, error) {
+			Enabled:     true,
+			Description: "",
+		}
+		if redactorEntry.Metadata.Name == "" {
+			newRedactorSpec.Name = slug
+			redactorEntry.Metadata.Name = slug
+		}
+	} else {
+		// unmarshal existing redactor, check if name changed
+		err = json.Unmarshal([]byte(redactString), &redactorEntry)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to parse redactor %s", slug)
+		}
+		redactorEntry.Metadata.Updated = time.Now()
+
+		if slug != getSlug(newRedactorSpec.Name) && newRedactorSpec.Name != "" {
+			// changing name
+			delete(configMap.Data, slug)
+			slug = getSlug(newRedactorSpec.Name)
+			redactorEntry.Metadata.Slug = slug
+			redactorEntry.Metadata.Name = newRedactorSpec.Name
+		}
+
+		if newRedactorSpec.Name == "" {
+			newRedactorSpec.Name = slug
+			redactorEntry.Metadata.Name = slug
+		}
+	}
+
+	redactorEntry.Redact = newRedactorSpec
+
+	jsonBytes, err := json.Marshal(redactorEntry)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to marshal redactor %s", slug)
+	}
+
+	configMap.Data[slug] = string(jsonBytes)
+
+	_, err = writeConfigmap(configMap)
+	if err != nil {
+		return nil, errors.Wrapf(err, "write configMap with updated redact")
+	}
+	return &redactorEntry.Metadata, nil
+}
+
+func DeleteRedact(slug string) error {
+	configMap, _, err := getConfigmap()
+	if err != nil {
+		return err
+	}
+
+	delete(configMap.Data, slug)
+
+	_, err = writeConfigmap(configMap)
+	if err != nil {
+		return errors.Wrapf(err, "write configMap with updated redact")
+	}
+	return nil
+}
+
+func getConfigmap() (*v1.ConfigMap, string, error) {
+	cfg, err := config.GetConfig()
+	if err != nil {
+		return nil, "failed to get cluster config", errors.Wrap(err, "failed to get cluster config")
+	}
+
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, "failed to create kubernetes clientset", errors.Wrap(err, "failed to create kubernetes clientset")
+	}
+
+	configMap, err := clientset.CoreV1().ConfigMaps(os.Getenv("POD_NAMESPACE")).Get("kotsadm-redact", metav1.GetOptions{})
+	if err != nil {
+		if !kuberneteserrors.IsNotFound(err) {
+			// not a not found error, so a real error
+			return nil, "failed to get kotsadm-redact configMap", errors.Wrap(err, "failed to get kotsadm-redact configMap")
+		} else {
+			// not found, so return empty string
+			return nil, "", nil
+		}
+	}
+	return configMap, "", nil
+}
+
+func writeConfigmap(configMap *v1.ConfigMap) (*v1.ConfigMap, error) {
+	cfg, err := config.GetConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get cluster config")
+	}
+
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create kubernetes clientset")
+	}
+
+	newConfigMap, err := clientset.CoreV1().ConfigMaps(os.Getenv("POD_NAMESPACE")).Update(configMap)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to update configmap")
+	}
+	return newConfigMap, nil
+}
+
+func getSlug(name string) string {
```

Comment:
I don't think this function is 1-to-1.  why is this needed?

Reply:
I thought there were restrictions on the keys a configmap could have in k8s
if there aren't, I'd love to drop slugs entirely